python/ray/worker.py

```diff
@@ -502,13 +511,20 @@ def submit_task(self, function_id, args, actor_id=None):
             # Look up the various function properties.
             function_properties = self.function_properties[
                 self.task_driver_id.id()][function_id.id()]
+            num_return_vals = function_properties.num_return_vals
```

Comment:
We can leave this for a subsequent PR depending on what you prefer, but I think it would be *much* cleaner to not have any special cases in `submit_task` at all and to just have the actor call `submit_task` differently. This would require the actor handle (as opposed to the worker object) to own the `actor_counter` and the `actor_dummy_object`, which I think makes a lot of sense.

I think we can also make it so that there are no special cases in `_process_task` and that the differences are handled inside of `actor_method_call`.

If you decide to defer this to a subsequent PR, can you submit an issue for this change?

Reply:
Can you describe a bit more how this would be done? I'm not sure if I see how we can call `submit_task` differently without changing its arguments, since it doesn't currently take a number of return values as an argument. Do you mean that we should do the increment to the number of return values when registering the function? I'm also not sure how we can have the actor add the dependency rather than the caller, since it seems like by the time the actor receives the task, the dependencies should already be decided?

Reply:
I suppose we could either increment the number of returns when registering the function or make `num_return_vals` an argument to `submit_task`.

And yeah, the caller needs to add the dependency. I didn't mean the actor itself would add the dependency but rather the actor "handle", which I'm using to refer to the actor object that the caller has (e.g., this is `NewClass` in `actor.py`). Does that make more sense?

Reply:
Yes, that does seem much cleaner. I pushed a commit to do this. Thanks!